### ISSUE TITLE
Lighting/shadows under water

### ISSUE THREAD

**Issue by tut-yury** [uji-ros-pkg/underwater_simulation#20]
Hi guys,
I am trying to simulate a dive inside a shipwreck, in real life it should be dark inside the ship with light coming form outside through holes (I use another model of a shipwreck, with actual hull).
In UWSim, however, darkness (using fog) depends only on distance from camera. If camera is inside the hull and the ship is 100 meters deep, it is still as bright as if it had no roof and a surface level...
Is there anything I can tweak in the world file, blend/ive file or in the code to make lighting/shadows work (better)?

**Comment by marioprats** [comment]
We could probably play with other light types with attenuation like point lights, and enabling shadows. Or maybe light attenuation could also be simulated at the shader level... I guess this is something where some research is needed, but definitely a good to have feature.

**Comment by perezsolerj** [comment]
Hi,

I worked on shadows while developing structured light project. As far as I know shaders do NOT compute shadows (only structured light shadows), but those projected shadows are really heavy (time consuming) require cameras and shadow precision depends on camera resolution. But maybe you can try creating a SLS Projector on light position and see what happens. Obviously this feature is quite different but it can be a nice start point. I will have a look at it as soon as posible.

Mario I think we can't just enable shadows because these option uses a shader and I think we would need to compose shaders as only one shader can be executed for each object.

I'm not sure how we should achieve this, but it will require changes in shaders. And, as Mario said, some research in shadowing techniques is probably needed.

**Comment by tut-yury** [comment]
I tried the SLS approach, but even with some hacking it "lights" not only the face of the surface/texture, but also the back side as well. Shadows also don't work too reliable, they are casted more-or-less correctly, when there is a bigger distance between objects (>2-3 metres), but surfaces closer than that behind the closest object dont' have shadows or have them incorrectly...
Below is an example of circular SLS light... as it can be seen ship's hull nor masts cast shadows...
![sls](https://f.cloud.github.com/assets/1423752/1516342/d7e986e8-4b12-11e3-8fe0-a173a6ca35ef.png)

**Comment by perezsolerj** [comment]
Hi Yury,

I had no time to write a good answer before, sorry. 

As I have no idea of your knowledge about shadow techniques I will explain how I did it on SLS, if you still know about it just skip it and got to the end!.

SLS shadows are created with a range camera that measures the distance from the SLS to the first object. Then when the shader is guessing if the place where the SLS is projected is a shadow or not checks that distance and compares it with theone being checked, so if its bigger it's a shadow.

The problem if you use this is you get self shadowing due to small errors. The result looks like this:

![captura de pantalla de 2013-11-21 17 58 55](https://f.cloud.github.com/assets/4101745/1593613/537a8dda-52ce-11e3-9790-9c71aee5ff4f.png)

So what it's usually done is add a small threshold. I choosed one that worked fine for short distances. As z-buffer  (camera distance measure) is not linear it seems to be failing on long distances, so just adjust on shader default_scene.frag, 0.001 worked fine for me.

// Used to lower moiré pattern and self-shadowing
shadowCoordinateWdivide.z -= 0.005;

Some minor changes in shader and the result is:

![captura de pantalla de 2013-11-21 18 12 27](https://f.cloud.github.com/assets/4101745/1593748/40d06158-52d0-11e3-870d-3689555bd36e.png)

Some weird shadows in the edges of the area covered by the SLS, but it looks promising!!!. Just changed following lines:

```
if( (texcolor.x+texcolor.y+texcolor.z)<2.0 && ShadowCoord.w > 0.0 && shadow==0.5 && texcolor!=vec4(1.0,1.0,1.0,1.0) && texcolor!=vec4(1.0,0.0,0.0,0.0)){
    textureColor = vec4( 0.0,0.0,0.0,1.0);
}
```

It can be improved a lot....
